## Profitable App Profiles for the App Store and Google Play Markets

Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

- [Google Play App dataset](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data of about 10000 (approx.) Android apps from Google Play.

- [App Store App dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) containing data of about 7000 (approx.) iOS apps from the App Store.


In [1]:
import csv

To make it easier to open and read multiple datasets in the required format, we'll write a function `open_and_read_dataset()`. The function returns a tuple that consists of the dataset header row and the complete dataset without the header row.

In [2]:
def open_and_read_dataset(filename):
    '''
    Input: 
    filename - name of .csv file to open
    
    Output: file contents as list of lists
    
    Description:
    Function to open a .csv file, read 
    each row/line of file as a list and
    store all the lists inside a list 
    '''
    file = open(filename, encoding='utf8')
    filecontents = csv.reader(file)
    dataset = list(filecontents)
    return dataset[0], dataset[1:]

Using the `open_and_read_dataset()`, we'll unpack the dataset header row and the complete dataset without the header row in separate variables.

In [3]:
googleplay_header, googleplay = open_and_read_dataset('googleplaystore.csv')
applestore_header, applestore = open_and_read_dataset('AppleStore.csv')

Next we'll write a function named `explore_data()` that'll print some rows from the dataset for exploring data rows in more readable way. The function also optionally prints the number of rows and columns the dataset holds.

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    '''
    Input: 
             dataset - dataset in the form of list of lists
                       (as returned by open_and_read_dataset() function)
               start - start index position for slicing the dataset
                 end - end index position for slicing the dataset
    rows_and_columns - set this flag to True only if the dataset being
                       passed doesn't have the header row
    
    Output:
    Prints the file lines as list from the list slice and print the
    number of rows and columns of the dataset if the dataset doesn't
    have the header row
    '''
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

**Google Play Android Apps Data**

In [5]:
print(googleplay_header)
print('\n')
google_data = explore_data(googleplay, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


We see that the `googleplaystore.csv` dataset has information of 10841 apps. Out of the 13 app attributes, the attributes that might be useful for the purpose of our analysis are `'App'`, `'Category'`, `'Reviews'`, `'Installs'`, `'Type'`, `'Price'`, and `'Genres'`.

**App Store iOS Apps Data**

In [6]:
print(applestore_header)
print('\n')
apple_data = explore_data(applestore, 0, 3, True)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 7197
Number of columns: 17


We see that the `AppleStore.csv` dataset has information of 10841 apps. Out of the 17 app attributes, the attributes that might be useful for the purpose of our analysis are `'track_name'`, `'currency'`, `'price'`, `'rating_count_tot'`, `'rating_count_ver'`, and `'prime_genre'`.

Before beginning our analysis, we need to make sure the data we analyze is accurate, otherwise the results of our analysis will be wrong. This means that we need to:

   - Detect inaccurate data, and correct or remove it.
   - Detect duplicate data, and remove the duplicates.

We want to perform the analysis on the apps that are free to download and install, and that are directed toward an English-speaking audience. This means that we'll need to:

   - Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
   - Remove apps that aren't free.

**Detect and Delete Incorrect/Inaccurate data**

In the `googleplaystore.csv` dataset, there is a row with incorrect value at index 10472. You can read more about this in the [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015#latest-600082) of this dataset on Kaggle.

- Let's print the row at index 10472

In [7]:
print(googleplay[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


- Let's print the header row along with the row that holds ***incorrect*** values

In [8]:
for field, data in zip(googleplay_header, googleplay[10472]):
    print('{:>14}: {}'.format(field, data))

           App: Life Made WI-Fi Touchscreen Photo Frame
      Category: 1.9
        Rating: 19
       Reviews: 3.0M
          Size: 1,000+
      Installs: Free
          Type: 0
         Price: Everyone
Content Rating: 
        Genres: February 11, 2018
  Last Updated: 1.0.19
   Current Ver: 4.0 and up


- Let's print the header row along with a row that holds ***correct*** values

In [9]:
for field, data in zip(googleplay_header, googleplay[0]):
    print('{:>14}: {}'.format(field, data))

           App: Photo Editor & Candy Camera & Grid & ScrapBook
      Category: ART_AND_DESIGN
        Rating: 4.1
       Reviews: 159
          Size: 19M
      Installs: 10,000+
          Type: Free
         Price: 0
Content Rating: Everyone
        Genres: Art & Design
  Last Updated: January 7, 2018
   Current Ver: 1.0.0
   Android Ver: 4.0.3 and up


By comparing the above two data row cells, we observe that the data for the app `Life Made WI-Fi Touchscreen Photo Frame` is jumbled up and not correctly ordered due to some missing data compared to the second cell with correct data values. Therefore we'll delete the row.

In [10]:
print('Number of rows: {}'.format(len(googleplay)))
del googleplay[10472]
print('Number of rows (after row deletion): {}'.format(len(googleplay)))

Number of rows: 10841
Number of rows (after row deletion): 10840


**Checking for Duplicate Rows**

Next, we explore the `googleplaystore.csv` dataset and try to detect duplicate rows if exists in the dataset.

- Let's check if the app `Instagram` appears more than once in the dataset 

In [11]:
for app_info_list in googleplay:
    app_name = app_info_list[0]
    if app_name == 'Instagram':
        print(app_info_list)
        print()

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']

['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']

['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']



From the above cell we see that there are 3 duplicate rows for the `'Instagram'` app.

We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

If you examine the rows we printed for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times.
We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

- Let's check the number of duplicate app entries in the dataset

In [12]:
duplicate_app_entries = []
unique_app_entries = []

for app_info_list in googleplay:
    app_name = app_info_list[0]
    if app_name in unique_app_entries:
        duplicate_app_entries.append(app_name)
    else:
        unique_app_entries.append(app_name)
        
num_duplicate_entries = len(duplicate_app_entries)
num_unique_entries = len(unique_app_entries)
print('Unique app entries in dataset: {}'.format(num_unique_entries))
print('Duplicate app entries in dataset: {}'.format(num_duplicate_entries))

Unique app entries in dataset: 9659
Duplicate app entries in dataset: 1181


We'll now create a dictionary to:
1. store the number of duplicate entries a particular app has
2. find the number apps with duplicate entries

In [13]:
duplicate_apps = {}
for app_name in duplicate_app_entries:
    if app_name not in duplicate_apps:
        duplicate_apps[app_name] = 1
    else:
        duplicate_apps[app_name] += 1
num_of_apps = len(duplicate_apps)
print('Number of apps with duplicate entries: {}'.format(num_of_apps))

Number of apps with duplicate entries: 798


- Let's print apps from the `duplicate_apps` dictionary along with the number duplicate entries.

In [14]:
# For pretty printing dictionaries for better readability
import json

In [15]:
print(json.dumps(duplicate_apps, indent=2))

{
  "Quick PDF Scanner + OCR FREE": 2,
  "Box": 2,
  "Google My Business": 2,
  "ZOOM Cloud Meetings": 1,
  "join.me - Simple Meetings": 2,
  "Zenefits": 1,
  "Google Ads": 2,
  "Slack": 2,
  "FreshBooks Classic": 1,
  "Insightly CRM": 1,
  "QuickBooks Accounting: Invoicing & Expenses": 2,
  "HipChat - Chat Built for Teams": 1,
  "Xero Accounting Software": 1,
  "MailChimp - Email, Marketing Automation": 1,
  "Crew - Free Messaging and Scheduling": 1,
  "Asana: organize team projects": 1,
  "Google Analytics": 1,
  "AdWords Express": 1,
  "Accounting App - Zoho Books": 1,
  "Invoice & Time Tracking - Zoho": 1,
  "Invoice 2go \u2014 Professional Invoices and Estimates": 1,
  "SignEasy | Sign and Fill PDF and other Documents": 1,
  "Genius Scan - PDF Scanner": 1,
  "Tiny Scanner - PDF Scanner App": 1,
  "Fast Scanner : Free PDF Scan": 1,
  "Mobile Doc Scanner (MDScan) Lite": 1,
  "TurboScan: scan documents and receipts in PDF": 1,
  "Tiny Scanner Pro: PDF Doc Scan": 1,
  "Docs To Go\u212

From the above cell, we understand that the app named `Quick PDF Scanner + OCR FREE` has 2 duplicate entries in the dataset and similarly for the other 797 apps. 

In [16]:
dataset_without_dup = len(googleplay) - num_duplicate_entries
print('Hence, our dataset must contain {} rows after removing {} duplicate entries from the dataset.'.format(dataset_without_dup, num_duplicate_entries))

Hence, our dataset must contain 9659 rows after removing 1181 duplicate entries from the dataset.


**Removing the Duplicate entries**

- We now know about the number of duplicate entries in the dataset, next we will remove the duplicate entries from the dataset by following the criteria that an app with duplicate entries having the highest value for the attribute/column`'Reviews'` (index 3) among the duplicate entries for that app will be kept in the dataset and remaining copies will be removed.



In [17]:
reviews_max = {}
for app_info_list in googleplay:
    app_name = app_info_list[0]
    n_app_review = float(app_info_list[3])
    # If app_name (key) exists in reviews_max dictionary and the value (number of reviews) 
    # coressponding to the key is less that the current value of the duplicate entry, then
    # update the lesser value with the current greater value
    if (app_name in reviews_max) and (reviews_max[app_name] < n_app_review):
        reviews_max[app_name] = n_app_review
    # If app_name (key) doesn't exist in reviews_max dictionary, then create an entry for
    # the key with the current value of n_app_review 
    elif app_name not in reviews_max:
        reviews_max[app_name] = n_app_review
    else:
        pass
print('The dictionary contains {} entries after removing {} duplicate entries which is the correct number of rows as expected.'.format(len(reviews_max), num_duplicate_entries))

The dictionary contains 9659 entries after removing 1181 duplicate entries which is the correct number of rows as expected.


Next, we'll create a dataset named `googleplay_clean` by selecting only the unique app data rows based on the criteria using the `reviews_max` dictionary that will hold all the unique data rows without any duplicates.

We will also use a list named `googleplay_added` which will just store app names that'll help us to keep track of apps that we already added.

In [18]:
googleplay_clean = []
googleplay_added = []

for app_info_list in googleplay:
    app_name = app_info_list[0]
    n_app_review = float(app_info_list[3])
    # We know that the reviews_max dictionary holds the duplicate entries in the form
    # where the app_name is mapped to the highest number_of_review for that app amongnst the
    # duplicate entries
    # So if the current app's number of review is equal to the max_number_of_review for that
    # app in the reviews_max dictionary and the app_name is not already added to the dataset
    # then append the app_info_list to the new dataset
    if (n_app_review == reviews_max[app_name]) and (app_name not in googleplay_added):
        googleplay_clean.append(app_info_list)
        googleplay_added.append(app_name)
print('The cleaned dataset without any duplicate data has {} rows'.format(len(googleplay_clean)))

The cleaned dataset without any duplicate data has 9659 rows


**Checking for Non-English Apps**

Recall that we want to perform the analysis on the apps that are free to download and install, and that are ***directed toward an English-speaking audience***. In the previous step, we managed to remove the duplicate app entries in the Google Play data set. If we explore the datasets further, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience.

- Let's look at some examples of non-english app data that exist in both the datasets

In the `googleplay_clean` dataset,

In [19]:
print(googleplay_clean[4412][0])
print()
print(googleplay_clean[7940][0])

中国語 AQリスニング

لعبة تقدر تربح DZ


And in the `applestore` dataset,

In [20]:
print(applestore[84][2])
print()
print(applestore[210][2])

同花顺-炒股、股票

央视影音-海量央视内容高清直播


Our next task is to remove all such non-english apps from both the datasets. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text. English text usually includes letters from the English alphabet, numbers composed of digits from `0 to 9`, punctuation marks `(., !, ?, ;),` and other symbols `(+, *, /)`.

Behind the scenes, each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character `'a' is 97`, character `'A' is 65`, and character `'爱' is 29,233`. We can get the corresponding number of each character using the `ord()` built-in function. The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the `ASCII (American Standard Code for Information Interchange)` system.

In [21]:
def app_name_eval(app_name):
    for character in app_name:
        # If an app name contains a character that is greater than 127, 
        # then it probably means that the app has a non-English name
        if ord(character) > 127:
            return False
    # If none of the characters of the app_name is greater that 127, 
    # then return True (i.ie., it is an english name)
    return True

- Let's check if the above function works correctly

In [22]:
result_1 = app_name_eval('Instagram')
print('English app name: {}'.format(result_1))

English app name: True


In [23]:
result_2 = app_name_eval('爱奇艺PPS -《欢乐颂2》电视剧热播')
print('English app name: {}'.format(result_2))

English app name: False


In [24]:
result_3 = app_name_eval('Docs To Go™ Free Office Suite')
print('English app name: {}'.format(result_3))

English app name: False


In [25]:
result_4 = app_name_eval('Instachat 😜')
print('English app name: {}'.format(result_4))

English app name: False


In all the four cases above, the function returns the correct result, but considering the last two cases, the app names can be considered as english names as only english letters are used but there are some special characters and emojis present whose ASCII value is more than 127.

In [26]:
print('{:>2}: {}'.format('😜', ord('😜')))
print('{:>3}: {}'.format('™', ord('™')))

 😜: 128540
  ™: 8482


If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English. Our filter function is still not perfect, but it should be fairly effective.

- Let's create a new function by changing the existing function `app_name_eval()`

In [27]:
def app_name_eval_updated(app_name):
    num_chars = 0
    for character in app_name:
        if ord(character) > 127:
            num_chars += 1
    if num_chars > 3:
        return False
    return True

- Let's check if the above function `app_name_eval_updated()` works correctly

In [28]:
result_5 = app_name_eval_updated('Docs To Go™ Free Office Suite')
print('English app name: {}'.format(result_5))

English app name: True


In [29]:
result_6 = app_name_eval_updated('Instachat 😜')
print('English app name: {}'.format(result_6))

English app name: True


In [30]:
result_7 = app_name_eval_updated('爱奇艺PPS -《欢乐颂2》电视剧热播')
print('English app name: {}'.format(result_7))

English app name: False


- The above function is working correctly as per the condition that an app is considered as an english named app if it contains more than 3 characters having a value greater than 127.

Next, we'll further clean both the datasets by removing the non-english apps

**Removing Non-English Apps**

In [31]:
def remove_non_eng_apps(dataset, appname_col_index):
    '''
    Input:
              dataset - dataset with english and non-english apps
                        (without header row)
    appname_col_index - app name column index in dataset
    
    Output:
      cleaned_dataset - list of lists containing info of only 
                        english apps
    '''
    cleaned_dataset = []
    for app_info_list in dataset:
        app_name = app_info_list[appname_col_index]
        if app_name_eval_updated(app_name):
            cleaned_dataset.append(app_info_list)
    return cleaned_dataset

In [32]:
googleplay_english = remove_non_eng_apps(googleplay_clean, 0)
applestore_english = remove_non_eng_apps(applestore, 2)

- Let's print some rows of both `googleplay_english` and `applestore_english` dataset

**Google Play Android Apps Data (english named apps)**

In [33]:
explore_data(googleplay_english, start=0, end=3, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


**App Store iOS Apps Data (english named apps)**

In [34]:
explore_data(applestore_english, start=0, end=3, rows_and_columns=True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 6183
Number of columns: 17


We see that, the `googleplay_english` dataset has data of 9614 english-named apps and the `applestore_english` dataset has data of 6183 english-named apps.

**Isolating free apps**

So far in the data cleaning process, we:
- Removed inaccurate data
- Removed duplicate app entries
- Removed non-English apps

Recall again that we want to perform the analysis on the apps that are ***free to download and install***, and that are directed toward an English-speaking audience. Our data sets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis. Isolating the free apps will be our last step in the data cleaning process.

In [35]:
def isolate_free_apps(dataset, price_col_index):
    '''
    Input:
            dataset - dataset with free and non-free english apps
                      (without header row)
    price_col_index - app price column index in dataset
    
    Output:
          free_apps - list of lists containing info of 
                      only free english apps
    '''
    free_apps = []
    for app_info_list in dataset:
        app_price = app_info_list[price_col_index]
        if app_price == '0':
            free_apps.append(app_info_list)
    return free_apps

In [36]:
googleplay_free_eng_apps = isolate_free_apps(googleplay_english, 7)
applestore_free_eng_apps = isolate_free_apps(applestore_english, 5)

- Let's explore both `googleplay_free_eng_apps` and `applestore_free_eng_apps` datasets

**Google Play Android Apps Data (free english named apps)**

In [37]:
explore_data(googleplay_free_eng_apps, start=0, end=3, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


**App Store iOS Apps Data (free english named apps)**

In [38]:
explore_data(applestore_free_eng_apps, start=0, end=3, rows_and_columns=True)

['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']


Number of rows: 3222
Number of columns: 17


After,
   - Removing inaccurate data
   - Removing duplicate app entries
   - Removing non-English apps
   - Isolating the free apps

We are now left with **8864 Android apps** and **3222 iOS apps** for analysis.

**Most Common Apps by Genre**

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

   - Build a minimal Android version of the app, and add it to Google Play.
   - If the app has a good response from users, we develop it further.
   - If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

- Let's analyze our app datasets and create a frequence table to determine some of the most common apps by genre

In [39]:
def frequency_table(dataset, genre_col_index):
    freq_table = {}
    total_rows = 0
    # Getting a total count of different types of app genres
    for app_info_list in dataset:
        total_rows += 1
        genre = app_info_list[genre_col_index]
        if genre not in freq_table:
            freq_table[genre] = 1
        else:
            freq_table[genre] += 1
    
    freq_table_percent = {}
    # Converting the total number into a percentage
    for genre in freq_table:
        percentage = (freq_table[genre] / total_rows)*100
        freq_table_percent[genre] = percentage
        
    return freq_table_percent

- The following function takes in the genre:count dictionary from the above function `frequency_table()` and then sorts the genre by counts in descending order.

In [40]:
def sort_freq_table(dataset, genre_col_index):
    freq_table_percentage = frequency_table(dataset, genre_col_index)
    list_of_tuples = []
    for genre in freq_table_percentage:
        genre_count_tuple = (freq_table_percentage[genre], genre)
        list_of_tuples.append(genre_count_tuple)
    # Sort in descending order
    sorted_genre_counts = sorted(list_of_tuples, reverse=True)
    for genre_tuple in sorted_genre_counts:
        print('{:>19}: {}'.format(genre_tuple[1],genre_tuple[0]))

**Google Play Android Apps Data**

In [41]:
print(googleplay_header,'\n\n')
category_col_index = googleplay_header.index('Category')
sort_freq_table(googleplay_free_eng_apps, category_col_index)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 


             FAMILY: 18.907942238267147
               GAME: 9.724729241877256
              TOOLS: 8.461191335740072
           BUSINESS: 4.591606498194946
          LIFESTYLE: 3.9034296028880866
       PRODUCTIVITY: 3.892148014440433
            FINANCE: 3.7003610108303246
            MEDICAL: 3.531137184115524
             SPORTS: 3.395758122743682
    PERSONALIZATION: 3.3167870036101084
      COMMUNICATION: 3.2378158844765346
 HEALTH_AND_FITNESS: 3.0798736462093865
        PHOTOGRAPHY: 2.944494584837545
 NEWS_AND_MAGAZINES: 2.7978339350180503
             SOCIAL: 2.6624548736462095
   TRAVEL_AND_LOCAL: 2.33528880866426
           SHOPPING: 2.2450361010830324
BOOKS_AND_REFERENCE: 2.1435018050541514
             DATING: 1.861462093862816
      VIDEO_PLAYERS: 1.7937725631768955
MAPS_AND_NAVIGATION: 1.3989169675090252
     FOOD_AND_D

We are not considering the `'Genres'` column from the `googleplay_free_eng_apps` dataset because the `'Genres'` column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

**App Store iOS Apps Data**

In [42]:
print(applestore_header,'\n\n')
prime_genre_col_index = applestore_header.index('prime_genre')
sort_freq_table(applestore_free_eng_apps, prime_genre_col_index)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 


              Games: 58.16263190564867
      Entertainment: 7.883302296710118
      Photo & Video: 4.9658597144630665
          Education: 3.662321539416512
  Social Networking: 3.2898820608317814
           Shopping: 2.60707635009311
          Utilities: 2.5139664804469275
             Sports: 2.1415270018621975
              Music: 2.0484171322160147
   Health & Fitness: 2.0173805090006205
       Productivity: 1.7380509000620732
          Lifestyle: 1.5828677839851024
               News: 1.3345747982619491
             Travel: 1.2414649286157666
            Finance: 1.1173184357541899
            Weather: 0.8690254500310366
       Food & Drink: 0.8069522036002483
          Reference: 0.5586592178770949
           Business: 0.5276225946617008
          

- Let's analyze the derived frequency tables and make observations.

**Google Play: Android Apps**

Q. **What are the most common genres?**

The top 5 genres are:

| Genre     | Frequency (in %)  |
| --------- | ----------------- |
| Family    | 18.90794223826714 |
| Game      | 9.724729241877256 |
| Tools     | 8.461191335740072 |
| Business  | 4.591606498194946 |
| Lifestyle | 3.903429602888086 |

Q. **What other patterns do you see?**

Looking at the top 10 genres, there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes.

**App Store: iOS Apps**

Q. **What are the most common genres?**

The top 5 genres are:

| Genre             | Frequency (in %)  |
| ----------------- | ----------------- |
| Games             | 58.16263190564867 |
| Entertainment     | 7.883302296710118 |
| Photo & Video     | 4.965859714463066 |
| Education         | 3.662321539416512 |
| Social Networking | 3.289882060831781 |

Q. **What other patterns do you see?**

Looking at the top 10 genres, AppStore is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare.

**Most popular Genre of app by Number of Users**

If there's a large number of apps for a particular genre, does that also imply that apps of that genre generally have a large number of users?

The frequency tables we analyzed on the previous screen showed us that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and fun apps. Now, we'd like to get an idea about the kind of apps with the most users.

One way to find out what genres have the most users is to calculate the average number of installs for each app genre.

**Google Play: Android Apps**

For the Google Play data set, we will use the `'Installs'` column to find which genre of apps have highest number of users. The `'Installs'` column values are not precise enough, instead most values are open-ended (100+, 1,000+, 5,000+, etc.)

We'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we'll need to convert each install number from string to float. This means we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error.

In [43]:
# Call the function to get the dictionary with unique genres (keys)
categories_android = frequency_table(googleplay_free_eng_apps, 1)
avgnum_installs = []
for genre_category in categories_android:
    total_installs = 0
    num_of_apps_in_genre = 0
    for app_info_list in googleplay_free_eng_apps:
        app_genre = app_info_list[1]
        if app_genre == genre_category:
            n_app_installs = app_info_list[5]
            n_app_installs = n_app_installs.replace(',','')
            n_app_installs = n_app_installs.replace('+','')
            n_app_installs = float(n_app_installs)
            total_installs += n_app_installs
            num_of_apps_in_genre += 1
    avg_num_of_installs = total_installs / num_of_apps_in_genre
    avgnum_installs.append((genre_category, avg_num_of_installs))

num_installs = []
for info_tuples in avgnum_installs:
    rev_tuple = (info_tuples[1], info_tuples[0])
    num_installs.append(rev_tuple)
sorted_num_installs = sorted(num_installs, reverse=True)
for genre_installs in sorted_num_installs:
    print('{:>19}: {}'.format(genre_installs[1], genre_installs[0]))

      COMMUNICATION: 38456119.167247385
      VIDEO_PLAYERS: 24727872.452830188
             SOCIAL: 23253652.127118643
        PHOTOGRAPHY: 17840110.40229885
       PRODUCTIVITY: 16787331.344927534
               GAME: 15588015.603248259
   TRAVEL_AND_LOCAL: 13984077.710144928
      ENTERTAINMENT: 11640705.88235294
              TOOLS: 10801391.298666667
 NEWS_AND_MAGAZINES: 9549178.467741935
BOOKS_AND_REFERENCE: 8767811.894736841
           SHOPPING: 7036877.311557789
    PERSONALIZATION: 5201482.6122448975
            WEATHER: 5074486.197183099
 HEALTH_AND_FITNESS: 4188821.9853479853
MAPS_AND_NAVIGATION: 4056941.7741935486
             FAMILY: 3695641.8198090694
             SPORTS: 3638640.1428571427
     ART_AND_DESIGN: 1986335.0877192982
     FOOD_AND_DRINK: 1924897.7363636363
          EDUCATION: 1833495.145631068
           BUSINESS: 1712290.1474201474
          LIFESTYLE: 1437816.2687861272
            FINANCE: 1387692.475609756
     HOUSE_AND_HOME: 1331540.5616438356
        

From the android apps data presented above, we see that apps from the **communication** genre have the highest number of average installs.

- Let's take a look at the apps that come under the `communication` genre and have number of installs within the range from 100,000,000+ to 1,000,000,000+.

In [44]:
for app_info_list in googleplay_free_eng_apps:
    if app_info_list[1] == 'COMMUNICATION' and (app_info_list[5] == '1,000,000,000+'
                                                or app_info_list[5] == '500,000,000+'
                                                or app_info_list[5] == '100,000,000+'):
        print('{:>50}: {}'.format(app_info_list[0],app_info_list[5]))

                                WhatsApp Messenger: 1,000,000,000+
                      imo beta free calls and text: 100,000,000+
                                  Android Messages: 100,000,000+
             Google Duo - High Quality Video Calls: 500,000,000+
          Messenger – Text and Video Chat for Free: 1,000,000,000+
                     imo free video calls and chat: 500,000,000+
                     Skype - free IM & video calls: 1,000,000,000+
                                               Who: 100,000,000+
        GO SMS Pro - Messenger, Free Themes, Emoji: 100,000,000+
                       LINE: Free Calls & Messages: 500,000,000+
                      Google Chrome: Fast & Secure: 1,000,000,000+
                    Firefox Browser fast & private: 100,000,000+
       UC Browser - Fast Download Private & Secure: 500,000,000+
                                             Gmail: 1,000,000,000+
                                          Hangouts: 1,000,000,000+
             

- Let's take a look at the apps that come under the runner-up `video-players` genre and have number of installs within the range from 100,000,000+ to 1,000,000,000+.

In [45]:
for app_info_list in googleplay_free_eng_apps:
    if app_info_list[1] == 'VIDEO_PLAYERS' and (app_info_list[5] == '1,000,000,000+'
                                                or app_info_list[5] == '500,000,000+'
                                                or app_info_list[5] == '100,000,000+'):
        print('{:>50}: {}'.format(app_info_list[0],app_info_list[5]))

                                           YouTube: 1,000,000,000+
                                  Motorola Gallery: 100,000,000+
                                   VLC for Android: 100,000,000+
                           Google Play Movies & TV: 1,000,000,000+
                                         MX Player: 500,000,000+
                                          Dubsmash: 100,000,000+
            VivaVideo - Video Editor & Photo Movie: 100,000,000+
VideoShow-Video Editor, Video Maker, Beauty Camera: 100,000,000+
                                 Motorola FM Radio: 100,000,000+


- Let's take a look at the apps that come under the `'social'` genre and have number of installs within the range from 100,000,000+ to 1,000,000,000+.

In [46]:
for app_info_list in googleplay_free_eng_apps:
    if app_info_list[1] == 'SOCIAL' and (app_info_list[5] == '1,000,000,000+'
                                                or app_info_list[5] == '500,000,000+'
                                                or app_info_list[5] == '100,000,000+'):
        print('{:>50}: {}'.format(app_info_list[0],app_info_list[5]))

                                          Facebook: 1,000,000,000+
                                     Facebook Lite: 500,000,000+
                                            Tumblr: 100,000,000+
                                         Pinterest: 100,000,000+
                                           Google+: 1,000,000,000+
                    Badoo - Free Chat & Dating App: 100,000,000+
                      Tango - Live Video Broadcast: 100,000,000+
                                         Instagram: 1,000,000,000+
                                          Snapchat: 500,000,000+
                                          LinkedIn: 100,000,000+
                    Tik Tok - including musical.ly: 100,000,000+
                           BIGO LIVE - Live Stream: 100,000,000+
                                                VK: 100,000,000+


From the above data on app installations in the `COMMUNICATION`, `VIDEO_PLAYERS` and `'SOCIAL'` genre, we observe that there are some leading app brands with over 1 billion installations and competing against such popular apps with new apps will be a difficult task until and unless the new app matches or is better at providing the services compared to the existing popular ones.

- Let's take a look at the apps that come under the `'PHOTOGRAPHY'` genre (4<sup>th</sup> on the most popular genres list) and have number of installs within the range from 100,000,000+ to 1,000,000,000+.

In [47]:
for app_info_list in googleplay_free_eng_apps:
    if app_info_list[1] == 'PHOTOGRAPHY' and (app_info_list[5] == '1,000,000,000+'
                                                or app_info_list[5] == '500,000,000+'
                                                or app_info_list[5] == '100,000,000+'):
        print('{:>55}: {}'.format(app_info_list[0],app_info_list[5]))

                          B612 - Beauty & Filter Camera: 100,000,000+
                 YouCam Makeup - Magic Selfie Makeovers: 100,000,000+
   Sweet Selfie - selfie camera, beauty cam, photo edit: 100,000,000+
                                          Google Photos: 1,000,000,000+
                                                Retrica: 100,000,000+
                                       Photo Editor Pro: 100,000,000+
         BeautyPlus - Easy Photo Editor & Selfie Camera: 100,000,000+
       PicsArt Photo Studio: Collage Maker & Pic Editor: 100,000,000+
                                   Photo Collage Editor: 100,000,000+
        Z Camera - Photo Editor, Beauty Selfie, Collage: 100,000,000+
     PhotoGrid: Video & Pic Collage Maker, Photo Editor: 100,000,000+
     Candy Camera - selfie, beauty camera, photo editor: 100,000,000+
                   YouCam Perfect - Selfie Photo Editor: 100,000,000+
      Camera360: Selfie Photo Editor with Funny Sticker: 100,000,000+
         S Photo E

- On observing the number of times the apps have been downloaded/installed in the `'PHOTOGRAPHY'` genre, we can see only one app has 1 billion+ download and hence we can say that there is a scope for a new better app in this genre in the Google Play Store.

- Photography apps are very famous among smartphone users and any app that provides descent editing tools and photo filters with good UI at the least can expect a good amount of growth in Google Play Store.  

**App Store: iOS Apps**

For the Google Play data set, we can find this information using the `'Installs'` column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `'rating_count_tot'`column.

In [48]:
prime_genres_ios = frequency_table(applestore_free_eng_apps, 12)
avgnum_reviews = []
for genre in prime_genres_ios:
    total_reviews = 0
    num_of_app_in_genre = 0
    for app_info_list in applestore_free_eng_apps:
        app_genre = app_info_list[12]
        if app_genre == genre:
            n_app_reviews = float(app_info_list[6]) # col: rating_count_tot
            total_reviews += n_app_reviews
            num_of_app_in_genre += 1
    avg_num_reviews = total_reviews / num_of_app_in_genre
    avgnum_reviews.append((avg_num_reviews, genre))

sorted_num_reviews = sorted(avgnum_reviews, reverse=True)
for genre_reviews in sorted_num_reviews:
    print('{:>19}: {}'.format(genre_reviews[1], genre_reviews[0]))

         Navigation: 86090.33333333333
          Reference: 74942.11111111111
  Social Networking: 71548.34905660378
              Music: 57326.530303030304
            Weather: 52279.892857142855
               Book: 39758.5
       Food & Drink: 33333.92307692308
            Finance: 31467.944444444445
      Photo & Video: 28441.54375
             Travel: 28243.8
           Shopping: 26919.690476190477
   Health & Fitness: 23298.015384615384
             Sports: 23008.898550724636
              Games: 22788.6696905016
               News: 21248.023255813954
       Productivity: 21028.410714285714
          Utilities: 18684.456790123455
          Lifestyle: 16485.764705882353
      Entertainment: 14029.830708661417
           Business: 7491.117647058823
          Education: 7003.983050847458
           Catalogs: 4004.0
            Medical: 612.0


- Let's take a look at the apps that come under the `'Navigation'` genre

In [49]:
for app_info_list in applestore_free_eng_apps:
    if app_info_list[12] == 'Navigation':
        print('{:>50}: {}'.format(app_info_list[2], app_info_list[6]))

   Waze - GPS Navigation, Maps & Real-time Traffic: 345046
                                       Geocaching®: 12811
  ImmobilienScout24: Real Estate Search in Germany: 187
                              Railway Route Search: 5
       CoPilot GPS – Car Navigation & Offline Maps: 3582
                Google Maps - Navigation & Transit: 154911


- Let's take a look at the apps that come under the runner-up `'Reference'` genre

In [50]:
for app_info_list in applestore_free_eng_apps:
    if app_info_list[12] == 'Reference':
        print('{:>100}: {}'.format(app_info_list[2],app_info_list[6]))

                                                                                               Bible: 985920
                                                               Dictionary.com Dictionary & Thesaurus: 200047
                                                      Dictionary.com Dictionary & Thesaurus for iPad: 54175
                                                  Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran: 18418
                                                                          Merriam-Webster Dictionary: 16849
                                                                                    Google Translate: 26786
                                                                                           Night Sky: 12122
                                                                                                WWDC: 762
                                                   Jishokun-Japanese English Dictionary & Translator: 0
                                

This genre `'Reference'` seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

**Conclusion**

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

- We recommend the `'PHOTOGRAPHY'` genre of apps for the Google Play Store.
- And we recommend the `'Reference'` genre of apps for the App Store.

If an app of this genre in the respective app markets gets really good response, we can then create a similar app for the other market.